src/wasm-stack.h

```diff
@@ -197,10 +197,21 @@ template<typename SubType> void BinaryenIRWriter<SubType>::write() {
   emitFunctionEnd();
 }
 
-// emits a node, but if it is a block with no name, emit a list of its contents
+// Emits a node in a position that can contain a list of contents, like an if
+// arm. This will emit the node, but if it is a block with no name, just emit
+// its contents. This is ok to do because a list of contents is ok in the wasm
+// binary format in such positions anyhow. When we read such code in Binaryen
+// we will end up creating a block for it (note that while doing so we create a
+// block without a name, since nothing branches to it, which makes it easy to
+// handle in optimization passes and when writing the binary out again).
 template<typename SubType>
 void BinaryenIRWriter<SubType>::visitPossibleBlockContents(Expression* curr) {
   auto* block = curr->dynCast<Block>();
+  // Even if the block has a name, check if the name is necessary (if it has no
+  // uses, it is equivalent to not having one). Scanning the children of the
+  // block means that this takes quadratic time, but it will be N^2 for a fairly
+  // small N since the number of nested non-block control flow structures tends
+  // to be very reasonable.
```

Comment:
I'm still not sure... Doesn't `BranchSeeker::has` mostly only check branches, or more precisely, name uses defined by `DELEGATE_FIELD_SCOPE_NAME_USE`? (Currently `br`, `br_table`, `br_on`, `rethrow`, and `try`(due to `delegate`) have this field)

> but it will be N^2 for a fairly small N since the number of nested non-block control flow structures tends to be very reasonable.

I'm not sure what do you mean by "non-block control flow structures". Do you mean branches here?

Also, it is N^2 because we need to check every child, and that N includes all children. We don't know whether it is a branch or not before checking it. So I'm also not sure what you mean by "a fairly small N".

I'm not suggesting the quadratic behavior is bad; it is preexisting anyway and it is for a reason  and as you said it doesn't cause unreasonable slowdown. I'm just not very good at understanding the comments.

Reply:
Non-block control flow structures are: If, Try, Loop. In each of their arms they call `visitPossibleBlockContents` (to try to avoid emitting a block there).

By small N I mean that there are few nested non-block control flow structures. This would be bad:
```wat
(if
  ..
  (if
    ..
    (if
      ..N such ifs..
      ..
        ..innermost child..
```
Those nested ifs lead to the innermost child being scanned N times where N is the number of those ifs (since each if scans all children). So the total time is `O(N*M)` if M is the total number of children of the first if. But, N is small in the real world, such if stacks are very rare (unlike block stacks).